presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -0,0 +1,206 @@
+                    }
+                    return projection;
+                })
+                .collect(toImmutableList());
+    }
+
+    public PageProcessorOutput process(ConnectorSession session, Page page)
+    {
+        // limit the scope of the dictionary ids to just one page
+        dictionarySourceIdFunction.reset();
+
+        if (page.getPositionCount() == 0) {
+            return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        }
+
+        if (filter.isPresent()) {
+            SelectedPositions selectedPositions = filter.get().filter(session, filter.get().getInputChannels().getInputChannels(page));
+            if (selectedPositions.isEmpty()) {
+                return EMPTY_PAGE_PROCESSOR_OUTPUT;
+            }
+
+            if (projections.isEmpty()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), Iterators.singletonIterator(new Page(selectedPositions.size())));
+            }
+
+            if (selectedPositions.size() != page.getPositionCount()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), new PositionsPageProcessorIterator(session, page, selectedPositions));
+            }
+        }
+
+        return new PageProcessorOutput(
+                page.getRetainedSizeInBytes(),
+                new PositionsPageProcessorIterator(session, page, SelectedPositions.positionsRange(0, page.getPositionCount())));
+    }
+
+    private class PositionsPageProcessorIterator
+            extends AbstractIterator<Page>
+    {
+        private final ConnectorSession session;
+        private final Page page;
+
+        private SelectedPositions selectedPositions;
+        private final Block[] previouslyComputedResults;
+
+        public PositionsPageProcessorIterator(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+        {
+            this.session = session;
+            this.page = page;
+            this.selectedPositions = selectedPositions;
+            this.previouslyComputedResults = new Block[projections.size()];
+        }
+
+        @Override
+        protected Page computeNext()
+        {
+            while (true) {
+                if (selectedPositions.isEmpty()) {
+                    return endOfData();
+                }
+
+                int batchSize = Math.min(selectedPositions.size(), projectBatchSize);
+                Optional<Page> result = processBatch(batchSize);
+
+                // if the page buffer filled up, so halve the batch size and retry
+                if (!result.isPresent()) {
+                    verify(batchSize > 1);
+                    projectBatchSize = projectBatchSize / 2;
+                    continue;
+                }
+
+                Page page = result.get();
+
+                // if we produced a large page, halve the batch size for the next call
+                long pageSize = page.getSizeInBytes();
+                if (page.getPositionCount() > 1 && pageSize > MAX_PAGE_SIZE_IN_BYTES) {
```

Comment:
this is inconsistent with: https://github.com/prestodb/presto/pull/7515/files/5ca5f9440235ef3822961e7ad8e273b32c1de522#diff-92486801dfdc982c80726599c9c31357R128. In https://github.com/prestodb/presto/pull/7515/files/5ca5f9440235ef3822961e7ad8e273b32c1de522#diff-92486801dfdc982c80726599c9c31357R128, we disallow too large pages, but here we allow it (and reduce `projectBatchSize`)